openedx/frontend-app-authoring
# feat: handle unsaved changes in text & problem editors

## Description

Fixes: https://github.com/openedx/frontend-app-authoring/issues/1388 and https://github.com/openedx/frontend-app-authoring/issues/1389.


https://github.com/user-attachments/assets/e6e93111-d715-481e-8043-4e47ca517b7e

- Which edX user roles will this change impact?: "Course Author".

## Supporting information

* `Private-ref`: [FAL-3919](https://tasks.opencraft.com/browse/FAL-3919)

## Testing instructions

* Edit any text or problem block from new mfe unit page as well as legacy unit page.
* Try going back or cancelling without any change, it should take you back without requiring any confirmation.
* Edit again and make some changes, now clicking on cancel button or close button or refreshing page or using browser back button should be blocked by a popup.
* Test cancelling and saving changes from the editor.

## Limitations

* The browsers ability to block users action is not perfect, for example, the event `beforeunload` which we use to detect navigation away from site does not trigger in mobile browsers in some cases. For more information please refer [this doc](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#usage_notes) about the event.
* This PR does not add the functionality of detecting unsaved chanegs in **video editor** yet as it seems to be bit more involved. **I can look into it if it is required.**
* React router has support for blocking navigation when required but it seems to refresh the page regardless of user clicking cancel or stay which keeps the user in the page but the changes are lost. So I had to use browser navigation to take user to editors even from new MFE unit page which can then make use of `beforeunload` event to block navigation when required. Not using react router to navigate to editor results in slow navigation although it is the default way of going back to unit page on clicking cancel button.
